test/asynchronous/test_database.py

```diff
@@ -0,0 +1,778 @@
+)
+
+from bson.codec_options import CodecOptions
+from bson.dbref import DBRef
+from bson.int64 import Int64
+from bson.objectid import ObjectId
+from bson.regex import Regex
+from bson.son import SON
+from pymongo import helpers_shared
+from pymongo.asynchronous import auth
+from pymongo.asynchronous.collection import AsyncCollection
+from pymongo.asynchronous.database import AsyncDatabase
+from pymongo.asynchronous.helpers import anext
+from pymongo.asynchronous.mongo_client import AsyncMongoClient
+from pymongo.errors import (
+    CollectionInvalid,
+    ExecutionTimeout,
+    InvalidName,
+    InvalidOperation,
+    OperationFailure,
+    WriteConcernError,
+)
+from pymongo.read_concern import ReadConcern
+from pymongo.read_preferences import ReadPreference
+from pymongo.write_concern import WriteConcern
+
+_IS_SYNC = False
+
+
+class TestDatabaseNoConnect(unittest.TestCase):
+    """Test Database features on a client that does not connect."""
+
+    client: AsyncMongoClient
+
+    @classmethod
+    def setUpClass(cls):
+        cls.client = AsyncMongoClient(connect=False)
+
+    def test_name(self):
+        self.assertRaises(TypeError, AsyncDatabase, self.client, 4)
+        self.assertRaises(InvalidName, AsyncDatabase, self.client, "my db")
+        self.assertRaises(InvalidName, AsyncDatabase, self.client, 'my"db')
+        self.assertRaises(InvalidName, AsyncDatabase, self.client, "my\x00db")
+        self.assertRaises(InvalidName, AsyncDatabase, self.client, "my\u0000db")
+        self.assertEqual("name", AsyncDatabase(self.client, "name").name)
+
+    def test_get_collection(self):
+        codec_options = CodecOptions(tz_aware=True)
+        write_concern = WriteConcern(w=2, j=True)
+        read_concern = ReadConcern("majority")
+        coll = self.client.pymongo_test.get_collection(
+            "foo", codec_options, ReadPreference.SECONDARY, write_concern, read_concern
+        )
+        self.assertEqual("foo", coll.name)
+        self.assertEqual(codec_options, coll.codec_options)
+        self.assertEqual(ReadPreference.SECONDARY, coll.read_preference)
+        self.assertEqual(write_concern, coll.write_concern)
+        self.assertEqual(read_concern, coll.read_concern)
+
+    def test_getattr(self):
+        db = self.client.pymongo_test
+        self.assertTrue(isinstance(db["_does_not_exist"], AsyncCollection))
+
+        with self.assertRaises(AttributeError) as context:
+            db._does_not_exist
+
+        # Message should be: "AttributeError: Database has no attribute
+        # '_does_not_exist'. To access the _does_not_exist collection,
+        # use database['_does_not_exist']".
+        self.assertIn("has no attribute '_does_not_exist'", str(context.exception))
+
+    def test_iteration(self):
+        db = self.client.pymongo_test
+        if "PyPy" in sys.version and sys.version_info < (3, 8, 15):
+            msg = "'NoneType' object is not callable"
+        else:
+            if _IS_SYNC:
```

Comment:
The synchronization process will take care of the renaming for us, this check is unnecessary.